### PR TITLE
SNOW-1637215: Add describe query telemetry 

### PR DESCRIPTION
**Pull request by sfc-gh-jdu**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1637215

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

  Add telemetry for describe query mentioned in https://docs.google.com/document/d/1DAN1I99BQd1CBBu_pLvWb9vLReTC7HIbfppxeVbXCSg/edit?tab=t.0#heading=h.r285g4qcslsz

### CODE REVIEW

File: src/snowflake/snowpark/_internal/analyzer/schema_utils.py [hunk 1]
```diff
@@ -90,7 +91,20 @@ def analyze_attributes(
             session._conn._cursor.description, session._conn.max_string_size
         )
 
-    return session._get_result_attributes(sql)
+    # collect describe query details for telemetry
+    stack = traceback.extract_stack(limit=10)[:-1]
+    if len(stack) >= 2:
+        stack_trace = [frame.line for frame in stack]
+    else:
+        stack_trace = None
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/analyzer/schema_utils.py:99]: are we doing this to figure out which is the biggest contributor of describe queries?

**Review by sfc-gh-aalam — APPROVED** [review]

File: src/snowflake/snowpark/_internal/analyzer/schema_utils.py (hunk 1 above)
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/analyzer/schema_utils.py:99]: yes

File: src/snowflake/snowpark/_internal/analyzer/metadata_utils.py [hunk 2]
```diff
@@ -17,6 +17,13 @@
     from snowflake.snowpark._internal.analyzer.analyzer import Analyzer
 
 
+class DescribeQueryTelemetryField(Enum):
+    TYPE_DESCRIBE_QUERY_DETAILS = "snowpark_describe_query_details"
+    SQL_TEXT = "sql_text"
+    E2E_TIME = "e2e_time"
+    STACK_TRACE = "stack_trace"
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/_internal/analyzer/metadata_utils.py:24]: It's good that we're capturing the stack trace. But I imagine we'll need to further process it to extract the specific operations resulting in describe queries?

**Review by sfc-gh-helmeleegy — APPROVED** [review]
LGTM, just had one question.

File: src/snowflake/snowpark/_internal/analyzer/metadata_utils.py (hunk 2 above)
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/analyzer/metadata_utils.py:24]: yea, we can do the parsing on the dashboard side

File: src/snowflake/snowpark/_internal/analyzer/schema_utils.py [hunk 3]
```diff
@@ -90,7 +91,20 @@ def analyze_attributes(
             session._conn._cursor.description, session._conn.max_string_size
         )
 
-    return session._get_result_attributes(sql)
+    # collect describe query details for telemetry
+    stack = traceback.extract_stack(limit=10)[:-1]
+    if len(stack) >= 2:
+        stack_trace = [frame.line for frame in stack]
+    else:
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/analyzer/schema_utils.py:98]: why when len(stack) < 2: the stack_trace is None?

Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/analyzer/schema_utils.py:98]: good catch, actually it should be `len(stack) > 0`

**Review by sfc-gh-yzou — APPROVED** [review]

**Pull request merged**